### PR TITLE
admin: Add support for prometheus summary metrics

### LINKED ISSUE #30471: Support prometheus summary metric types on the admin endpoint
*Support prometheus summary metric types on the admin endpoint*: *The prometheus stats admin endpoint currently supports histograms based on the statically configured buckets, but doesn't support quantile-based summaries*

*Description*:
The prometheus stats endpoint currently only returns histogram metrics for the configured buckets. A potentially more useful way of reporting these would be to use the histogram quantiles and report as a prometheus "summary" metric type. There is some prior art in controlling how histograms are reported (https://github.com/envoyproxy/envoy/pull/25812), which could be supported with a query param.

Current behavior:
```
~ > curl http://127.0.0.1:9901/stats/prometheus
# TYPE envoy_server_initialization_time_ms histogram
envoy_server_initialization_time_ms_bucket{le="0.5"} 0
envoy_server_initialization_time_ms_bucket{le="1"} 0
envoy_server_initialization_time_ms_bucket{le="5"} 0
...
envoy_server_initialization_time_ms_bucket{le="1800000"} 1
envoy_server_initialization_time_ms_bucket{le="3600000"} 1
envoy_server_initialization_time_ms_bucket{le="+Inf"} 1
envoy_server_initialization_time_ms_sum{} 165
envoy_server_initialization_time_ms_count{} 1
```

Intended Change:
```
~ > curl http://127.0.0.1:9901/stats/prometheus?histogram_emit_mode=summary
# TYPE envoy_server_initialization_time_ms summary
envoy_server_initialization_time_ms{quantile="0"} 160
envoy_server_initialization_time_ms{quantile="0.25"} 162.5
envoy_server_initialization_time_ms{quantile="0.5"} 165
...
envoy_server_initialization_time_ms{quantile="0.995"} 169.95
envoy_server_initialization_time_ms{quantile="0.999"} 169.99
envoy_server_initialization_time_ms{quantile="1"} 170
envoy_server_initialization_time_ms_sum{} 165
envoy_server_initialization_time_ms_count{} 1
```

### PR DESCRIPTION
**Pull request by andybradshaw**
Commit Message: Add support for prometheus summary metrics on the admin endpoint
Additional Description: Adds support emitting prometheus "summary" metrics for the internal histogram quantiles by supplying a query parameter. Multiple modes are supported, as in https://github.com/envoyproxy/envoy/pull/25812, and can be either `histogram`, `summary`, or `histogram,summary`. 
Risk Level: Low, no changes to existing default behavior
Testing: Added unit tests for histogram, summary, and summary+histogram emission
Docs Changes: Added documentation to the admin home page, and to the published admin docs around an optional query parameter.
Release Notes: Added a note in the `small_feature` section.

Fixes #30471

### CODE REVIEW

File: source/server/admin/prometheus_stats.cc
Comment (ggreenway) [inline source/server/admin/prometheus_stats.cc:319]: If you emit both types, I believe this will currently have duplicates for `_sum` and `_count`, which is invalid in the exposition format. I think it's probably not worth supporting both at the same time; just make it either summary or histograms.

File: docs/root/operations/admin.rst
Comment (ggreenway) [inline docs/root/operations/admin.rst:775]: I think `histogram mode` is a more concise name.

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

File: docs/root/operations/admin.rst
Comment (jmarantz) [reply docs/root/operations/admin.rst:775]: +1

`histogram_mode` I assume though?

File: source/server/admin/admin.cc
Comment (jmarantz) [inline source/server/admin/admin.cc:226]: consider also 'detail' which is a new mode available in the admin UI and the histogram API, showing the full detail of what the underlying histogram detail contains. The disadvantage of that relative to the "cumulative" or "disjoint" is that multiple Envoys may choose different buckets based on the data they see, so Prom might not be able to aggregate them.

Moreover, it's a little confusing that you are defining a new histogram enum that's a lot like the one `/stats` already has: 
```
  Admin::ParamDescriptor histogram_buckets{Admin::ParamDescriptor::Type::Enum,
                                           "histogram_buckets",
                                           "Histogram bucket display mode",
                                           {"cumulative", "disjoint", "detailed", "none"}};
```
see source/server/admin/stats_handler.cc:180 which is similar to this but different. And what's super confusing is that /stats?format=prometheus basically does an internal redirect to /stats/prometheus. See source/server/admin/stats_handler.cc:82. Should we just re-use that param-descriptor so this is all compatible across formats?

In that case we'd just use `histogram_buckets_mode` for consistency with what's in the product now.

File: source/server/admin/stats_params.h
Comment (jmarantz) [inline source/server/admin/stats_params.h:72]: it's definitely unexpected to see these 2 fields in this structure, and the second one only applies to prom.

File: source/server/admin/admin.cc
Comment (jmarantz) [reply source/server/admin/admin.cc:226]: I want to point out also that `"none"` above really means `"summary"` and is the default setting for that field.

It occurs to me know that the name `"none"` is misleading (this was my fault for not thinking that thorugh) and we should really rename that `"summary"` and let `"none"` be a permanently deprecated synonym.

Comment (jmarantz) [reply source/server/admin/admin.cc:226]: and one thing that's a little annoying, when merging these overlapping params, is that the default for histogram _buckets_mode" for non-prom stats is `"none"` (aka 'summary') and the default I guess for Prom is effectively `cumulative` I think.

Comment (jmarantz) [reply source/server/admin/admin.cc:226]: I thought about this a little more and I don't see an issue. If unspecified we can give prom a different behavior than non-prom (text,html,json) but they should still be the same set of options.

Comment (andybradshaw) [reply source/server/admin/admin.cc]: Had gone through a similar thought process initially around whether to just use `none`, but decided to err on the side of not changing existing defaults/behavior 😅 I've updated the PR to reflect what I believe is the desired behavior; only emit summary metrics on an explicitly set `histogram_buckets=none` query param.

It feels a little off to be checking things about `params.query_` in the prometheus handler, but that seemed to be the least invasive way to get the desired behavior. An alternative would be to add an explicit `Utility::HistogramBucketsMode::Unset` variant, and then different handlers can have different default behaviors, although I'm not sure if that would be even more confusing to users.

File: changelogs/current.yaml
Comment (jmarantz) [inline changelogs/current.yaml:55]: I would vote for renaming "none" to "summary" in this PR, leaving 'none' as a synonym for "summary".

File: docs/root/operations/admin.rst
Comment (jmarantz) [inline docs/root/operations/admin.rst:773]: For text and json, all 4 histogram_buckets modes are supported, and currently for prom, that is ignored, and in this PR you are adding support for 'summary'.

Right now you don't have 'detailed' or 'cumulative' supported for Prom and I'd consider that tech-debt; we don't have to solve it in this PR but we should leave TODOs. Actually TBH I'm not sure why 'cumulative' is ever useful but someone wanted it in the past and added it.

And we should document exactly what happens if the user specifies a not-yet-supported option like prometheus/detailed or prometheus/cumulative.

WDYT?

File: docs/root/operations/admin.rst
Comment (jmarantz) [inline docs/root/operations/admin.rst:776]: this doc does not have as much detail about what you'll get in Prom format compared to the various options for text or json.

File: source/server/admin/prometheus_stats.cc
Comment (jmarantz) [inline source/server/admin/prometheus_stats.cc:321]: I think it'd be nicer to put all the semantics of parsing the query and using a different default bucket-mode into stats_params.cc rather than re-parsing it here.

File: source/server/admin/stats_params.cc
Comment (jmarantz) [inline source/server/admin/stats_params.cc:3]: `#include "source/common/server/admin/utils.h"` I think.

File: test/server/admin/prometheus_stats_test.cc
Comment (jmarantz) [inline test/server/admin/prometheus_stats_test.cc:1]: remove?

**Comment by jmarantz** [comment]
needs main merge also

/wait

File: source/server/admin/admin.cc
Comment (jmarantz) [reply source/server/admin/admin.cc]: agreed -- let's centralize the handling of the inconsistent defaults in stats_params.cc instead.

File: docs/root/operations/admin.rst
Comment (ggreenway) [reply docs/root/operations/admin.rst:773]: The prometheus exposition format defines exactly how histograms are expressed. The only configurable/changeable behavior is the published buckets. No other options should be supported on prometheus histograms. (With the exception of switching to summaries instead of histograms; that is also well-defined in the exposition format)

File: test/server/admin/stats_html_render_test.cc
Comment (jmarantz) [inline test/server/admin/stats_html_render_test.cc:41]: Can you factor out the body of this test to a helper function so we can share the code between HistogramUnset and HistogramSummary? I think we should also test the compatibility of "none" in a third call to that helper.

**Review by jmarantz — COMMENTED** [review]
lgtm at high level with a few minor issues.

File: source/server/admin/prometheus_stats.cc
Comment (andybradshaw) [reply source/server/admin/prometheus_stats.cc:377]: The switch bit makes sense, and while I think that returning an error on unsupported modes makes sense, unfortunately that would be a behavioral break, since this currently works:
```
~  curl -vv http://127.0.0.1:9901/stats/prometheus?histogram_buckets=detailed >/dev/null
*   Trying 127.0.0.1:9901...

> GET /stats/prometheus?histogram_buckets=detailed HTTP/1.1
> Host: 127.0.0.1:9901
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< content-type: text/plain; charset=UTF-8
< cache-control: no-cache, max-age=0
< x-content-type-options: nosniff
< date: Fri, 03 Nov 2023 20:26:24 GMT
< server: envoy
< transfer-encoding: chunked
<
```

Defer to you all in terms of how much of a "break" this really is, since it seems like this was mostly unintentionally supported before.

File: test/server/admin/stats_html_render_test.cc
Comment (andybradshaw) [reply test/server/admin/stats_html_render_test.cc:41]: I'm not sure how much this would actually buy us, as the tests are pretty slim as is. Also, I don't think testing "none" is important here, since the HTML renderer should only care about the possible bucket modes, not the query params.

File: source/server/admin/prometheus_stats.cc
Comment (jmarantz) [reply source/server/admin/prometheus_stats.cc:377]: My preference is to fail requests for disjoint/detailed even though the buckets-mode is ignored now. @ggreenway WDYT?

Also I think we need to add the histogram_buckets param to source/server/admin/admin.cc in the block below
```
          makeHandler("/stats/prometheus", "print server stats in prometheus format",
```
and you can just give the 2 legal options for histogram_buckets there, patterning after "/stats" handler declaration. That won't stop someone from sending a request, bypassing the UI, with an invalid bucket-mode so you'll have to check that on entry to the prom handler.

Note that /stats?format=prometheus works and redirects to  /stats/prometheus handling.

File: test/server/admin/stats_html_render_test.cc
Comment (jmarantz) [reply test/server/admin/stats_html_render_test.cc:41]: I think factoring out the test body makes it more obvious that the Unset and Summary have the same behavior, without having to read and compare a few lines of numbers.

But you are right about not being able to test 'none' because this test specifies an enum, not a query-param, and you have the parsing test covered in utils_test.cc.

File: source/server/admin/prometheus_stats.cc
Comment (jmarantz) [reply source/server/admin/prometheus_stats.cc:377]: note this is still pending but we can wait for Greg to weigh in.

**Review by jmarantz — COMMENTED** [review]
/wait

File: source/server/admin/prometheus_stats.cc
Comment (ggreenway) [reply source/server/admin/prometheus_stats.cc:377]: I'm fine with breaking requests with an invalid bucket type request, even though it's currently ignored.

**Comment by jmarantz** [comment]
2 AIs to get this PR merged as I see it.

1. make unexpected buckets yield a 400 and test (required) 
2. factor out the common parts of the unit test per comment (preferred but I won't hold up the PR for it)

**Comment by andybradshaw** [comment]
Apologies for the delay on this. I've attempted to use the `IS_ENVOY_BUG` approach, but I'm not sure that does what we want here, since it seems like it kills the process (although that might only be in tests?? new to both C++ and this project, so it's entirely possible I misinterpreted your earlier comment 😅).

**Review by ggreenway — DISMISSED** [review]
/wait

**Comment by ggreenway** [comment]
Also, please fix merge conflicts

**Comment by andybradshaw** [comment]
Apologies for letting this fester, not sure if I can re-open from pushing/commenting

File: source/server/admin/prometheus_stats.cc
Comment (andybradshaw) [reply source/server/admin/prometheus_stats.cc:385]: Updated to add a `validateParams`, which seemed like the least invasive way to make this change. I tried a few other approaches, but each seemed to have their own problems... would be curious if I'm missing somthing:

#### Option 1. Make `statsAsPrometheus` return an `absl::StatusOr<uint64_t>`
This seemed like a good way to tie validation and emission of all the bucket mode cases in one place, but causes a lot of churn in tests/unrelated things. Additionally, we'd need to move histogram handling to the top of the function to prevent any writes making it to the response before we error out, which feels really weird.

#### Option 2. Parse `Utility::HistogramBucketsMode` into `Utility::PrometheusHistogramBucketsMode`
This change would still give us the benefit of there only being one code path that operates/validates the provided bucket mode, and each switch can be fully exhaustive/not have empty behavior as exists right now. This would be a bit more plumbing, and I'm not sure is the direction you'd want to go with this in the future.

In any case, this now returns a 400 and an error message on invalid bucket modes:
```
~ > curl -vv http://127.0.0.1:9901/stats/prometheus?histogram_buckets=disjoint
* Uses proxy env variable no_proxy == 'cluster.local'
*   Trying 127.0.0.1:9901...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9901 (#0)
> GET /stats/prometheus?histogram_buckets=disjoint HTTP/1.1
> Host: 127.0.0.1:9901
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 400 Bad Request
< content-type: text/plain; charset=UTF-8
< cache-control: no-cache, max-age=0
< x-content-type-options: nosniff
< date: Mon, 08 Jan 2024 21:20:52 GMT
< server: envoy
< transfer-encoding: chunked
<
* Connection #0 to host 127.0.0.1 left intact
unsupported Prometheus histogram bucket mode
```

**Review by jmarantz — DISMISSED** [review]
Thank you!

File: source/server/admin/prometheus_stats.cc
Comment (jmarantz) [inline source/server/admin/prometheus_stats.cc:328]: stupid gcc build gives this error:
```
2024-01-09T15:46:32.6877730Z source/server/admin/prometheus_stats.cc: In static member function ‘static absl::Status Envoy::Server::PrometheusStatsFormatter::validateParams(const Envoy::Server::StatsParams&)’:
2024-01-09T15:46:32.6880770Z source/server/admin/prometheus_stats.cc:329:1: error: control reaches end of non-void function [-Werror=return-type]
```

So annoying; the code is correct. But as a workaround I suggest making absl::Status a temp and set that temp in the switch statement, fall off the end, and return it.

**Review by jmarantz — COMMENTED** [review]
/wait

**Review by jmarantz — APPROVED** [review]

**Comment by jmarantz** [comment]
@ggreenway i'm good with this if you want to take a final pass

**Comment by jmarantz** [comment]
needs main merge.

@ggreenway you want to do a final senior maintainer pass?

**Comment by ggreenway** [comment]
@andybradshaw merge conflict on the changelog. I'm reviewing the rest of the code concurrently.

/wait

File: source/server/admin/prometheus_stats.cc
Comment (ggreenway) [inline source/server/admin/prometheus_stats.cc:296]: I'm not sure what the default formatting for a double is for `{3}` here. Do you think some specific precision should be specified?

File: source/server/admin/prometheus_stats.cc
Comment (ggreenway) [inline source/server/admin/prometheus_stats.cc:289]: Given that a summary can't be aggregated in the same way a histogram can, would it make more sense to use `intervalStatistics()` here? https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#summary suggests that the last 5-10 minutes of data are preferable; we don't have any way to get that time window though with the collected data, at least not as a rolling window.

File: source/server/admin/prometheus_stats.cc
Comment (ggreenway) [inline source/server/admin/prometheus_stats.cc:297]: The spec says that `NaN` is an allowed value (https://prometheus.io/docs/instrumenting/exposition_formats/). Should we leave it as-is here instead of translating it to 0?

**Review by ggreenway — CHANGES_REQUESTED** [review]

File: source/server/admin/prometheus_stats.cc
Comment (andybradshaw) [reply source/server/admin/prometheus_stats.cc]: Yeah, that's probably a good idea, although I'm a little confused by [this comment](https://github.com/envoyproxy/envoy/blob/3164f06df3269e5041f1a6f0e650bfbe94f25653/source/server/admin/prometheus_stats.cc#L112-L116) about fixed-point not being supported, since it seems pretty well documented in the [fmt library](https://hackingcpp.com/cpp/libs/fmt.html#fp-numbers)? Not sure what a sane default precision should be... first thought was to use `.32g` like in the bucket output above.

File: source/server/admin/prometheus_stats.cc
Comment (andybradshaw) [reply source/server/admin/prometheus_stats.cc]: Yeah, this is a bit unfortunate, as I would imagine getting a rolling window would be quite a bit more work...

Another approach could be to use tags to emit both the interval and cumulative summaries as the `quantileSummary` function does?

File: source/server/admin/prometheus_stats.cc
Comment (ggreenway) [reply source/server/admin/prometheus_stats.cc]: Honestly, not sure about the fmtlib `f` format vs that comment. I'm pretty sure I wrote the envoy comment. But I don't recall the situation. Probably either that I completely missed `f` in the docs, or it didn't work in the way we needed (maybe not enough significant digits or something?). 

I'd say for now use the existing `.32g` format, and someone should circle back in a separate change and see if we can switch them all to something better or not.

File: source/server/admin/prometheus_stats.cc
Comment (ggreenway) [reply source/server/admin/prometheus_stats.cc]: I'm not sure if prometheus (the server/scraper) would ingest everything properly with two sets of quantiles, only different in tags. I'm not sure what the right thing to do here is. Agreed that getting the rolling window would be a lot more work.

File: docs/root/operations/admin.rst
Comment (ggreenway) [inline docs/root/operations/admin.rst:777]: I think it's worth documenting that each emitted summary is over the interval of the last stats flush interval (and link to `stats_flush_interval` docs).

**Review by ggreenway — DISMISSED** [review]
Overall I think this is looking pretty good. One additional doc improvement. 

@jmarantz do you want to take another look also?

**Review by jmarantz — APPROVED** [review]

**Comment by jmarantz** [comment]
@andybradshaw CI is being a little flaky I think. It says 'presubmit' failed but I couldn't find an actual error.

Can you merge main to restart it?

**Pull request merged**
